Experimentelle geführte Erstellung der Konfigurationsdatei

DerAndereAndi opened the pull request:
Dies ist eine **experimentelle** Umsetzung im folgenden aufgeführten Ziele. Es kann sein dass dies irgendwann wieder entfernt wird!
Diese Umsetzung soll auch beim Lernen helfen wie eine anwenderfreundliche Konfiguration aussehen sollte/kann/muss.

## Ziel dieses PRs:

- Eine geführte Erstellung der Konfigurationsdatei
- Direktes Testen ob die Konfiguration jedes Gerätes auch funktioniert
- Konfigurationsabhängigkeiten durch direkte Konfiguration der Abhängigkeit lösen (z.B. Sponsorshipt required)
- Konfiguration in für Anwender bekannte Informationen und möglichst wenig Informationseingabe zu ermöglichen, z.B. anhand von Produkten anstatt aus der Implementierungssicht
- Bisherige Konfiguration funktioniert weiter

## Was ist implementiert

- Templates für bekannte Geräte, diese müssen überprüft und ggfs. aktualisiert werden
- Geführte Konfiguration in deutsch
- Es werden Meters, Vehicles, Chargers und Loadpoints konfiguriert. Aber nicht in all derer Möglichenkeiten, sondern das notwendige

## Was steht mittelfristig an, wenn der Weg passt

- Unterstützung für weitere Konfigurationsbereiche
- Generierte Dokumentation der unterstützen Geräte aus Anwendersicht (nicht wie heute die Implementierungssicht) 
- Bearbeiten der Konfiguration (nur nach dem neuen Schema)
- Saubere Implementierung damit auch eine Web-GUI ermöglicht werden kann
- Web-basierte Konfiguration inkl. Bearbeitung
- Unterstützung für mehrere Sprachen, z.B. englisch

## Wie es verwendet

- Start über `evcc configure`
- Die Konfiguration wird in `evcc.yaml` geschrieben. Falls diese existiert kann ein alternativer Dateiname angegeben werden

andig committed:
Add draft infrastrucuture for providing complex templates

andig committed:
wip

andig committed:
wip

andig committed:
Update sunspec.yaml

DerAndereAndi committed:
Include any yaml template in the folder

Also added another example

andig committed:
Add rendering sample from params and remove Params from rendering yaml

andig committed:
remove sample

andig committed:
Improve sample rendering

andig committed:
Render all usage types

andig committed:
Render usages

andig committed:
wip

andig committed:
Create /templates folder

DerAndereAndi committed:
Fix build

DerAndereAndi committed:
Fix sprig version

andig committed:
Trim code

andig committed:
Add tests

andig committed:
Better test logging

andig committed:
Parallel tests

DerAndereAndi committed:
Add modbus template support

- templates can have the magic comment `# ::modbus-setup::` to define that this is the place for modbus specific parameters
- `params.name: modbus` defines the possible modbus interfaces via `choice: ["rs485", "tcpip"]`
- valid `modbus` values for `evcc.config` are `rs485serial`, `rs485tcpip`, `tcpip`
- Includes more examples for `E3DC`, `Eastron SDM`, `Fronius Gen24 Plus`, `SMA Inverter`, `SMA Meter`, `Solarlog`

andig committed:
Fix tests

andig committed:
add acceptable errors

andig committed:
Add code generator for proxies

DerAndereAndi committed:
Use funk.ContainsString

DerAndereAndi committed:
Partly fix build

It still panics as there are no templates for `charger` and `vehicle` which means those subdirectories under `docs` are missing

andig committed:
Add test property to supplement default

andig committed:
Remove hints and empty default values

andig committed:
Add docs and split modbus handling

andig committed:
Simplify

andig committed:
remove hint

DerAndereAndi committed:
Fix proxy and generate issues with modbus

andig committed:
wip

DerAndereAndi committed:
Fix generate

DerAndereAndi committed:
Update templates

- Add description for usage in CLI configure
- Params have value for usage in CLI configure
- RednerProxyWithValues to render CLI configured templates

DerAndereAndi committed:
Add first implementation of CLI configure task

Invoke via `evcc configure`

DerAndereAndi committed:
Add eebus setup to generated config

andig committed:
Partial merge of config ui code for reuse

andig committed:
wip

andig committed:
Add vehicle config

andig committed:
wip

andig committed:
Export vehicle config

DerAndereAndi committed:
Allow adding multiple battery meters

DerAndereAndi committed:
Automatically assign device names

DerAndereAndi committed:
Actually ask for battery meters

DerAndereAndi committed:
Add template requirements

Allow to define other items to be required for this device to work, e.g. if it is based on EEBUS and thus requires the EEBUS server be setup and running

DerAndereAndi committed:
Merge branch 'master' into demo/params

# Conflicts:
#	vehicle/nissan.go

DerAndereAndi committed:
Fix proxy rendering with templates

DerAndereAndi committed:
Add more working template options

- `mask` for masking user input, e.g. passwords
- `advanced` for defining params the user does not need to be asked (optionals, that have a `default` set) to reduce complexity
- `example` for defining example values, e..g `WVW...` vins, that are not `default` values, because they won't work if they are not provided
- add required check

DerAndereAndi committed:
Remove params from rendering with no values

Remove template params that don't have a default and no value from being rendered

Otherwise optional values that are provided with no values (e.g. vin) would be rendered as `vin:`

DerAndereAndi committed:
Fix batteries list config param name

DerAndereAndi committed:
Update sma template

DerAndereAndi committed:
Merge branch 'master' into demo/params

DerAndereAndi committed:
Add sample charger and vehicle template support

DerAndereAndi committed:
Align template type names

DerAndereAndi committed:
More adjustments

DerAndereAndi committed:
Some more loadpoint additions

- Show charger description and charger config number when describing that the loadpoint for that charger is now to be done
- Show  the vehicle title (or description) when asking if it will charge at this loadpoint
- Added question for the maximum power the charger can provide to derive `mincurrent` and `maxcurrent` from that
- Added question for default charging mode
- Added check if charger provides meter data
- Ask for a separate charge meter if charger provides no meter data

DerAndereAndi committed:
Move charger setup into loadpoint setup

DerAndereAndi committed:
More additions

- Option to validate input data for int values
- Added option to specify max amperage and amount of phases manually

DerAndereAndi committed:
Remove evcc.yaml dependency from configure

DerAndereAndi committed:
Add localization support

DerAndereAndi committed:
Hide proxy debug output

DerAndereAndi committed:
Merge branch 'master' into demo/params

DerAndereAndi committed:
Undo vehicle auto proxy generation changes

DerAndereAndi committed:
Fix docs template generation

- Example values weren't used
- Modbus wasn't used if no usage was defined

DerAndereAndi committed:
Only use example values only for docs

DerAndereAndi committed:
Remove generated yaml in docs from gitignore

DerAndereAndi committed:
Add generated doc yaml files

DerAndereAndi committed:
Translation into German

DerAndereAndi committed:
Save the configuration into a file

DerAndereAndi committed:
Fix Wallbe template ports

DerAndereAndi committed:
Fix wallbe template type

DerAndereAndi committed:
Add missing space

DerAndereAndi committed:
Add sponsorship requirement support

DerAndereAndi committed:
Update PromptUI to v0.9

DerAndereAndi committed:
Fix template

DerAndereAndi committed:
Move PromptUI code into separate file

DerAndereAndi committed:
Replace PromptUI with Survey

DerAndereAndi committed:
Make sure passwords are always strings

DerAndereAndi committed:
Add hint to manual installation on cancel

DerAndereAndi committed:
Add cfos devices

DerAndereAndi committed:
Add vzlogger templates

DerAndereAndi committed:
Make sure string values are used in hyphens

DerAndereAndi committed:
Updat label, help, example and default usage

DerAndereAndi committed:
Fix last wallbe template

DerAndereAndi committed:
Some fixes

DerAndereAndi committed:
Another vzlogger fix

DerAndereAndi committed:
Bugfix and more meters

DerAndereAndi committed:
Add support for all in one setups

Now the first step is to ask the user for a meter device that provides multiple meters, e.g. grid & PV

DerAndereAndi committed:
VZLogger template fixes

DerAndereAndi committed:
Add support for linked devices

E.g. Kostal Plenticore always have a Kostal Smart Energy Meter. Only used when chosen as a "Komplettsystem"

DerAndereAndi committed:
Minor refactoring

DerAndereAndi committed:
Some function name changes

DerAndereAndi committed:
Add experimental warning

DerAndereAndi committed:
Fix typo

DerAndereAndi committed:
Fix a crash

DerAndereAndi committed:
Actually test the vehicle data

DerAndereAndi committed:
Multiple updates

- restructure and renamend singlesetup into guidedsetup on root level
- split SMA HM and EM into separate devices
- option to define hardware requirements
- fixed input validation
- added boolean type
- typos
- various other minor changes

DerAndereAndi committed:
Charge mode in german

DerAndereAndi committed:
Fix input validation

DerAndereAndi committed:
Minor vin help text improvement

DerAndereAndi committed:
First step in refactoring

DerAndereAndi committed:
More refactoring

DerAndereAndi committed:
Fix " encoding

This still results in issues if the value is a string but the yaml encoder doesn't like it because of missing ""

DerAndereAndi committed:
Fix device testing

DerAndereAndi committed:
Fix multiple usages in guided setup

DerAndereAndi committed:
Minor output style fix

DerAndereAndi committed:
Make sure string values are quoted

DerAndereAndi committed:
Set default value types & simplify checks

DerAndereAndi committed:
Minor text addition

DerAndereAndi committed:
Minor template update

DerAndereAndi committed:
More updates

- Added support for floats and according checks
- Updated texts

DerAndereAndi committed:
Fix setting default ValueType

DerAndereAndi committed:
Clarify different VW template descriptions

DerAndereAndi committed:
Merge branch 'master' into feature/cli-configure

DerAndereAndi committed:
Fix batteries reference

DerAndereAndi committed:
Remove old generated docs

DerAndereAndi committed:
Add missing line feed at eof

DerAndereAndi committed:
Merge branch 'master' into feature/cli-configure

DerAndereAndi committed:
Add template testing

- Senec Meter test is still failing
- PSA Vehicle tests are still failing

DerAndereAndi committed:
Fix meter tests

- Added LG login error
- Fixed Senec template

DerAndereAndi committed:
Disable temple doc generation for now

DerAndereAndi committed:
Fix linter warnings

DerAndereAndi committed:
Merge branch 'master' into feature/cli-configure

DerAndereAndi committed:
Minor change

DerAndereAndi committed:
Add docs generation for templates

DerAndereAndi committed:
Localization updates

DerAndereAndi committed:
Add Description to generated doc

DerAndereAndi committed:
Move more text into localized file

DerAndereAndi committed:
Fix guided setup not using proper usage

DerAndereAndi committed:
Fix linter error and minor change

DerAndereAndi committed:
Localization fixes

DerAndereAndi committed:
Text polishing

DerAndereAndi committed:
Update FrtizDECT Template

DerAndereAndi committed:
Improve fritz template

DerAndereAndi committed:
Update generated doc files

DerAndereAndi committed:
Minor refactor

DerAndereAndi committed:
Add multi language support for templates

DerAndereAndi committed:
Minor german text updates

DerAndereAndi committed:
Added english localization

DerAndereAndi committed:
Add missing localization string

DerAndereAndi committed:
Fix linter warning

DerAndereAndi committed:
Fix doc generation

DerAndereAndi committed:
Update templates

DerAndereAndi committed:
Merge branch 'master' into feature/cli-configure

# Conflicts:
#	go.mod
#	go.sum
#	meter/config_test.go

DerAndereAndi committed:
Fix merge bug

DerAndereAndi committed:
Added known test failure

andig commented on charger/proxy.go:
```diff
@@ -0,0 +1,51 @@
+		// println("")
+
+		// render the proxy
+		_, err := tmpl.RenderProxy()
```
warum müssen wir alle Proxies beim Start rendern? Wär das nicht ein Test?

DerAndereAndi replied:
müssen wir nicht. ist noch ein Überbleibsel wie die Kommentare von deinem Code. Im Test ist das ja schon drin. Mache ich alles hier raus.

andig commented on cmd/configure.go:
```diff
@@ -0,0 +1,38 @@
+func runConfigure(cmd *cobra.Command, args []string) {
+	impl := &configure.CmdConfigure{}
+
+	lang := ""
```
var lang string

andig commented on cmd/configure.go:
```diff
@@ -0,0 +1,38 @@
+		lang = langFlag
+	}
+
+	logLevel := ""
```
Warum nicht wie bei den anderen Commands? Gleiche Lösungen vermeiden doppelte Probleme.

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,230 @@
+)
+
+func (c *CmdConfigure) configureDeviceGuidedSetup() {
+	var repeat bool = true
```
repeat := true

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,230 @@
+
+	deviceItem := device{}
+
+	for ok := true; ok; ok = repeat {
```
das kann man bestimmt schöner schreiben

DerAndereAndi replied:
hast ne Idee wie?

DerAndereAndi replied:
Habs geändert.

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,230 @@
+
+		usageFound, usageChoices := c.paramChoiceValues(templateItem.Params, templates.ParamUsage)
+		if !usageFound {
+			fmt.Println("ERROR: Device template is missing valid usages!")
```
ist das nicht ein panic und müsste vorher in einem Test abgefangen werden?

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,230 @@
+			values := c.processConfig(linkedTemplateItem.Params, category, false)
+			deviceItem, err := c.processDeviceValues(values, linkedTemplateItem, deviceItem, category)
+			if err != nil {
+				if err != c.errDeviceNotValid {
```
errors.Is

?

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,230 @@
+	case DeviceClassVehicle:
+		v, err = vehicle.NewFromConfig(instance.Type, instance.Other)
+	}
+	if err != nil {
```
return v, err

andig commented on cmd/configure/helper.go:
```diff
@@ -0,0 +1,312 @@
+		}
+	}
+
+	return nameFound, choices
```
ist nameFound nicht einfach len(choices)>0? Wozu brauchts Beides?

DerAndereAndi replied:
stimmt

andig commented on cmd/configure/main.go:
```diff
@@ -0,0 +1,276 @@
+
+	bundle := i18n.NewBundle(language.German)
+	bundle.RegisterUnmarshalFunc("toml", toml.Unmarshal)
+	_, err := bundle.ParseMessageFileBytes([]byte(lang_de), "localization/de.toml")
```
if _, err := ....; err != nil

andig commented on cmd/configure/main.go:
```diff
@@ -0,0 +1,276 @@
+
+	fmt.Println()
+
+	filename := "evcc.yaml"
```
Globale const?

andig commented on meter/config_test.go:
```diff
@@ -40,3 +43,55 @@ func TestMeters(t *testing.T) {
 	}
 }
+
+func TestProxyMeters(t *testing.T) {
```
können wir das generell in ein proxy_test auslagern und den config_test unangetastet lassen?

DerAndereAndi replied:
und die Acceptables dann auch auslagern?

andig replied:
Gerne lassen wo die waren

DerAndereAndi replied:
aehm. dann muss ich sie im proxy_test duplizieren, nicht schön

andig replied:
Ich meinte in der datei, aber zugreifbar

andig commented on meter/config_test.go:
```diff
@@ -40,3 +43,55 @@ func TestMeters(t *testing.T) {
+		values := tmpl.Defaults(true)
+
+		// Modbus default test values
+		if values["modbus"] != nil {
```
alle map keys sollten definierte Konstanten sein

andig commented on templates/template_modbus.go:
```diff
@@ -0,0 +1,104 @@
+//go:embed modbus.tpl
+var modbusTmpl string
+
+func (t Template) RenderModbusTemplate(values map[string]interface{}, indentlength int) (string, error) {
```
Der Teil sieht hacky aus. Was soll hier passieren?

DerAndereAndi replied:
Hier werden die, je nach wert von `modbus`(also `rs485serial`, `rs485tcpip` oder `tcpip`), passenden modbus spezifischen Parameter, die der Benutzer angegeben hat, an der "richtigen" Stelle (Einrücken) in der Hierarchie der Konfiguration hinzugefügt. Damit das rendern und das instanziieren dann auch funktioniert.

andig replied:
puh. Könnten wir das nicht durch eingebundene Templates "schöner" lösen?

DerAndereAndi replied:
Das wird in den Templates ja mit dem magic comment gelöst. Und damit hast diese ganzen Cases nicht x-fach in vielen Templates drin und damit auch Fehleranfälligkeit.

DerAndereAndi replied:
Beispiel: https://github.com/evcc-io/evcc/pull/1888/files#diff-a8d84805f72971c964a1331f6aac5fbaf867ac015bc9eaac7140a94c24f8ce99

DerAndereAndi replied:
Du müsstest den Inhalt von modbus.tpl dann an jede Stelle in jedes Template reinkopieren. Imho nicht der richtige Weg

andig replied:
Du kannst Templates in Templates includieren- ich hatte eher in die Richtung gedacht.

DerAndereAndi replied:
Auch da müssen dann die entsprechenden Variablen gesetzt werden. Der Code bleibt imho fast identisch.

DerAndereAndi replied:
Erledigt, danke für den Hint.

DerAndereAndi committed:
Use `template:` instead of `type: t_`

Templates are now identified by a `template` property instead dof reusing the `type` property and having the need for a prefix

DerAndereAndi committed:
Add missing localization

DerAndereAndi committed:
Code review related changes

DerAndereAndi committed:
Run missing generated docs

DerAndereAndi committed:
Split up tests

DerAndereAndi committed:
Some more cleanup

andig commented on cmd/config.go:
```diff
@@ -156,6 +161,10 @@ func (cp *ConfigProvider) configureChargers(conf config) error {
 		}
 
+		// Support for template definitions
+		if cc.Template != "" {
```
mhhhhm- brauche ich das hier oder können wir das nach NewFromConfig verschieben? Letzteres wäre deutlich schönere Kapselung.

andig commented on cmd/config.go:
```diff
@@ -50,8 +51,8 @@ func (conf *mqttConfig) RootTopic() string {
 type qualifiedConfig struct {
-	Name, Type string
-	Other      map[string]interface{} `mapstructure:",remain"`
+	Name, Type, Template string
```
Siehe nächster Kommentar

DerAndereAndi replied:
Welcher nächster Kommentar? Da steht "s.o.". Möchte mich ungern hier im Kreis zwischen den Kommentaren drehen :D

andig replied:
Ich meine wir brauchen das Template erst in NewfromConfig, aber nicht hier! Das leakt Implementierungsdetails die hier irrelevant sind.

DerAndereAndi replied:
aehm. und wie kommt dann `template: sma-hm` aus der Konfig hier rein? Wenn es da nicht drin ist, kann ich das in der Konfig Datei nicht verwenden

DerAndereAndi replied:
In der Konfigurationsdatei steht ja so etwas:

```
meters:
- template: sma-home-manager
  usage: grid 
  host: 192.168.1.2
```

Damit das funktioniert, muss `Template` hier in den struct rein. Nochmal, in der Konfig soll so etwas knappes / kurzes stehen, nicht das komplizierte was es ersetzt.

andig replied:
Rein ja, hier nein. Das gehört in newtemplatefromconfug, da muss nochmal ein decode passieren um das template vom rest zu trennen. Dass es template gibt ist erst und nur dort relevant.

andig replied:
Das passt schon alles, erscheint mir aber noch zu global umgesetzt zu sein

DerAndereAndi replied:
Done

andig commented on cmd/config.go:
```diff
@@ -179,6 +188,10 @@ func (cp *ConfigProvider) configureVehicles(conf config) error {
 			return fmt.Errorf("cannot create %s vehicle: missing name", humanize.Ordinal(id+1))
 		}
 
+		// Support for template definitions
```
s.o.

andig commented on cmd/configure.go:
```diff
@@ -0,0 +1,36 @@
+		lang = langFlag
+	}
+
+	util.LogLevel(viper.GetString("log"), nil)
```
yay :)

andig commented on cmd/configure.go:
```diff
@@ -0,0 +1,36 @@
+	impl := &configure.CmdConfigure{}
+
+	var lang string
+	langFlag, err := cmd.Flags().GetString("lang")
```
Direkt zuweisen und panic falls err?

andig commented on cmd/configure/configure.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+type loadpoint struct {
+	Title       string
```
Perspektivisch können wir was aus core wiederverwenden, TODO für später

DerAndereAndi replied:
Füge das als kommentar in den code hinzu

andig commented on cmd/configure/configure.go:
```diff
@@ -0,0 +1,132 @@
+	Chargers   []device
+	Vehicles   []device
+	Loadpoints []loadpoint
+	Site       struct {
```
dito TODO

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+func (c *CmdConfigure) configureLinkedTypes(templateItem templates.Template) {
+	var repeat bool = true
```
dito oben

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,227 @@
+	}
+}
+
+func (c *CmdConfigure) configureDeviceCategory(deviceCategory DeviceCategory) (device, error) {
```
Aussagekräftige Kommentare für alle Funktionen wären hilfreich fürs Verständnis

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,227 @@
+		v, err = vehicle.NewFromConfig(instance.Type, instance.Other)
+	}
+
+	return v, err
```
yay :)

andig commented on cmd/configure/helper.go:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	err := c.processDeviceRequirements(templateItem)
+	if err != nil {
```
hier bitte auch einfacher

andig commented on cmd/configure/helper.go:
```diff
@@ -0,0 +1,312 @@
+		}
+		sponsortoken := c.askValue(question{
+			label:    c.localizedString("Requirements_Sponsorship_Token_Input", nil),
+			help:     "",
```
weg

andig commented on cmd/configure/helper.go:
```diff
@@ -0,0 +1,312 @@
+	if templateItem.Requirements.Eebus {
+		if c.configuration.EEBUS() == "" {
+			eebusConfig, err := c.eebusCertificate()
+
```
Leerzeile weg

DerAndereAndi committed:
Make sure naming is correct in template

DerAndereAndi committed:
Move template and type handling into NewFromConfig

DerAndereAndi commented on cmd/configure/helper.go:
```diff
@@ -0,0 +1,312 @@
+		return false
+	}
+
+	for _, choice := range choices {
```
choices hab ich nicht als Type `[]string`  sondern `[]DeviceCategory`

DerAndereAndi committed:
More review changes

DerAndereAndi committed:
Another review change

DerAndereAndi committed:
Fix template tests

DerAndereAndi committed:
Fix charger tests

DerAndereAndi committed:
Switch to nested templates

Remove hacky "magic comment" implementation

DerAndereAndi committed:
Minor fixes

DerAndereAndi committed:
Add template to json schema

DerAndereAndi committed:
Don't use `template` as a `type` replacement

andig commented on charger/config.go:
```diff
@@ -28,6 +29,9 @@ var registry chargerRegistry = make(map[string]func(map[string]interface{}) (api
 
 // NewFromConfig creates charger from configuration
 func NewFromConfig(typ string, other map[string]interface{}) (v api.Charger, err error) {
+	if typ == "template" {
```
Dann kannst Du das hier auch entsorgen. Einfach template als neuen Typ registrieren!

andig commented on charger/proxy.go:
```diff
@@ -0,0 +1,34 @@
+	"gopkg.in/yaml.v3"
+)
+
+func init() {
```
Weg damit. Templates erst bei Bedarf erzeugen! Spart startupzeit und Speicher.

andig commented on cmd/configure/configure.go:
```diff
@@ -0,0 +1,132 @@
+	c.config.SponsorToken = token
+}
+
+func (c *Configure) EEBUS() string {
```
Brauchst Du egrl für jeden getter eine Funktion oder tuts nicht auch der direkte zugriff aufs Property?

andig commented on cmd/configure/eebus.go:
```diff
@@ -0,0 +1,58 @@
+	details := communication.ManufacturerDetails{
+		DeviceName:    "EVCC",
+		DeviceCode:    "EVCC_HEMS_01",
+		DeviceAddress: "EVCC_HEMS",
```
Sind die Konstanten nicht von irgendwoher dupliziert? Original benutzen? Oder direkt die Details aus einer Funktion holen?

DerAndereAndi committed:
Add todo comments

DerAndereAndi committed:
Use properties directly

DerAndereAndi committed:
Reuse eebus manufacturer details

DerAndereAndi committed:
Optimize template usage

- register template for charger, meter, vehicle
- only instantiate configured templates

DerAndereAndi committed:
More comments

DerAndereAndi committed:
Use const instead of string value

DerAndereAndi committed:
Fix a bug with default value types

DerAndereAndi committed:
Add option to define modbus defaults

Devices may have specific modbus defaults for baudrate and comset, these can now be set unter `param: name: modbus` as `baudrate` and `comset`

DerAndereAndi commented:
Es wird so langsam rund. 🙇

andig commented on charger/proxy.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+func NewChargerFromTemplateConfig(other map[string]interface{}) (api.Charger, error) {
+	name := other["template"].(string)
```
ich würde wieder mapstructure nehmen- immer gleiche Lösungen für gleiches Problem!

DerAndereAndi replied:
Hast einen Tip wie die anderen Parameter außer `template` beim dekodieren ignoriert werden können? Hier ist nur `template` relevant, der Rest ist ja bei jedem Template anders

DerAndereAndi replied:
Ich bräuchte wohl ein `util.DecodeOther` aber mit `ErrorUnused: false`

andig replied:
Genau- den Rest in einen Sammler schieben:

```
type qualifiedConfig struct {
	Name, Type string
	Other      map[string]interface{} `mapstructure:",remain"`
}
```

Der Trick ist das `mapstructure:",remain"`

andig commented on templates/init.go:
```diff
@@ -0,0 +1,68 @@
+)
+
+//go:generate go run generate/generate.go
+func init() {
```
das macht eine Menge Yaml parsen beim Programmstart. Können wir das bitte nicht tun oder zur Not wenn gar nix besseres einfällt durch ein

    sync.Once.Do()

kapseln?

DerAndereAndi replied:
Das war dein Code für das Generieren der Docs, hab da noch nichts mit gemacht. Der ist ja nur für go:generate notwendig, sonst nicht. Wie willst das haben?

andig replied:
dann gerne ins generate. aktuell wird er immer aufgerufen.

DerAndereAndi replied:
mom, da wird die Liste der Templates geholt, damit die dann auch instantiiert werden können. Ne File-liste sollte genügen, und dann das instantiieren wird ja eh geparst

DerAndereAndi replied:
Hab mal `sync.Once` hinzugefügt. Parsing weglassen geht nicht, weil sonst die Templates nicht über den Namen sondern nur über den Dateinamen des Templates identifiziert werden könnten. Kann man machen, dann stehen die halt in den Konfigurationen drin.  Was ist dir lieber?

andig replied:
dann lassen wir's mal so

andig commented on templates/vehicle/kia.yaml:
```diff
@@ -0,0 +1,19 @@
+template: kia
+description: Kia
+params:
```
Könnten wir die "dummen" Parameter nicht mit irgendeinem Includemechanismus, z.B. Golang Templates, hier rein bauen? Das ist in fast jedem Auto x-fach wiederholt und wenn die Identifiers x-fach dazu kommen gibts genau so oft auch Anpassungen.

Dopplengen vermeiden!

DerAndereAndi replied:
Was sind für dich denn "dumme" Parameter? Meinst du `title`, `user`, `password`, `vin`, `capacity`? Also die am häufigsten auftreten und dann 1 Mal definiert werden. Und für die Ausnahmen (z.b. evnotify) dann halt nicht nehmen.

andig replied:
Genau, da kann identifiers gleich mit rein

andig replied:
da Du dann 2x templates parsen musst würde es sich anbieten, andere Klammerung zu nehmen (bei go template einstellbar), einmal für "prepare" und einmal für "render". Sonst kommen die sich ins Gehege

DerAndereAndi replied:
Über templates wurde es echt komplex. Habe jetzt `paramsbase:` hinzugefügt, da ist momentan nur `vehicles` definiert. In `Params` kann man die dann ergänzen oder von bestehenden `Default` oder `Example` anpassen.

DerAndereAndi committed:
Review related changes

DerAndereAndi committed:
Make sure parsing is only done once

andig commented on templates/init.go:
```diff
@@ -24,33 +27,35 @@ const (
-		if d.IsDir() {
-			return nil
-		}
+	templatesOnce.Do(func() {
```
äh.... du machst das 1x, aber immer noch im init, also IMMER. So bringt das nix ;)

DerAndereAndi replied:
Ja. Ändere es nachher. War in Eile.

DerAndereAndi replied:
Habs nochmal auf eine andere Art gemacht.

DerAndereAndi committed:
Only load templates on demand per class

DerAndereAndi committed:
Fix lint error

DerAndereAndi committed:
Show a message when testing is not possible

DerAndereAndi committed:
Allow adding devices even when test failed

DerAndereAndi committed:
Fix typo

DerAndereAndi committed:
Add `paramsbase`for templates

This allows to predefine a set of params, but also allows to add more in the template and overwrite `default` or `example` values of predefined params

DerAndereAndi committed:
Improved handling of failed configurations

DerAndereAndi committed:
Minor style updates

DerAndereAndi committed:
Fix configuration file space inconsistency

DerAndereAndi committed:
Fix another typo

DerAndereAndi committed:
Refactor device testing

DerAndereAndi committed:
Fix overwriting and extending params

DerAndereAndi committed:
Create proper yaml files with proper quoting

Big thanks to @andig !!!

DerAndereAndi committed:
Fix showing proper device title when added

DerAndereAndi committed:
Add optional advanced render mode

This renders the configuration, so it can be used in builds not supporting the new tempaltes

DerAndereAndi committed:
Add support for sma hems

DerAndereAndi committed:
Fix typo

DerAndereAndi committed:
Add support for loglevels

We need to update the templates and add the corresponding `loglevel`

DerAndereAndi committed:
Change advanced mode to expand & fix missing name

DerAndereAndi committed:
Add a few more loglevels to tempaltes

DerAndereAndi committed:
Improve texts when adding another

naltatis commented on templates/template.go:
```diff
@@ -0,0 +1,299 @@
+	values = t.ModbusValues(values)
+
+	for item, p := range values {
+		values[item] = yamlQuote(p.(string))
```
Hier knallts wenn in einer Template-Config im yaml eine Zahl steht.

Beispiel:
```
vehicles:
- type: template
  template: tesla
  title: foobar
  capacity: 82     <<<
```

oder

```
meters:
- type: template
  template: vzlogger_http
  usage: grid
  host: 192.168.31.53
  port: 9000             <<<<
```

DerAndereAndi replied:
Danke, gefixt.

naltatis replied:
Klappt. Danke!

DerAndereAndi committed:
Fix templates

naltatis commented on templates/meter/solaredge.yaml:
```diff
@@ -0,0 +1,61 @@
+    id: 1
+    subdevice: 1 # Metering device
+    value: 203:W
+    scale: -1
```
Das funktioniert so nicht mit meinem Wechselrichter (SolarEdge SE8K, LCD, Bj 2017) so leider nicht.

Meine aktuelle und funktionierende Config sieht so aus:
```
Das `timeout` wird bei mir unbedingt benötigt da das standard Timeout nicht ausreicht.
  - name: solaredge
    type: modbus
    uri: 192.168.31.149:502
    id: 1
    model: sunspec
    timeout: 3s
```
Das `subdevice` verursacht bei mir bspw. Fehler:
```
   Error: cannot create meter 'custom': power: sunspec: subdevice 1 not found
```

DerAndereAndi replied:
`Timeout` habe ich dem template hinzugefügt. Was mach ich dann mit `subdevice`?

naltatis replied:
War meine Verwirrung. Das Subdevice steht nur beim Grid drin. Was mein WR gar nicht hat weil kein Hybrid. Der PV Part im Template funktioniert für mich.

DerAndereAndi committed:
Fix rendering with non strings

DerAndereAndi committed:
Add timeout to solaredge template

DerAndereAndi committed:
Set sponsortoken so tests don't fail

DerAndereAndi committed:
Remove templates and generates docs from PR

DerAndereAndi committed:
Reorganize code directory structure

DerAndereAndi committed:
Add docs  generation to go:generate

DerAndereAndi committed:
Add missing embed file

DerAndereAndi commented:
@andig damit der Build nicht mehr failed, müssen wir den neuen PR mit den templates in main mergen, und main dann hier rein. Soll ich?

andig commented:
> @andig damit der Build nicht mehr failed, müssen wir den neuen PR mit den templates in main mergen, und main dann hier rein. Soll ich?

Trau Dich :)

DerAndereAndi committed:
Merge branch 'master' into feature/cli-configure

andig committed:
Minor cleanup

andig commented on cmd/configure/devicetest.go:
```diff
@@ -0,0 +1,139 @@
+type DeviceTestResult string
+
+const (
+	DeviceTestResult_Valid              DeviceTestResult = "Valid"
```
die _ sind immer wenn es übersetzte Strings sind? Sonst in Go eher unüblich, find ich aber zur Unterscheidung gut.

DerAndereAndi replied:
nein, wird nicht übersetzt. Habs umbenannt

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+				fmt.Println()
+				fmt.Println(err)
+			}
+			fmt.Println()
```
hoch ziehen und nur 1x println

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+	c.configureLinkedTypes(templateItem)
+}
+
+// let the user configure devices that are marked as being linked to a guided device
```
// configureLinkedTypes ...

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+
+	device := device{
+		Name:     DeviceCategories[deviceCategory].defaultName,
+		Title:    "",
```
Leere Inits weg

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+		Yaml:     "",
+	}
+
+	deviceDescription := ""
```
var ... string

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+
+	deviceDescription := ""
+
+	for ok := true; ok; {
```
?

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+				fmt.Println()
+				fmt.Println(err)
+			}
+			fmt.Println()
```
hoch schieben

andig replied:
stimmt :O

andig commented on cmd/configure/flow.go:
```diff
@@ -0,0 +1,198 @@
+
+	c.configuration.AddDevice(device, deviceCategory)
+
+	deviceTitle := ""
```
var ...

andig commented on cmd/configure/helper.go:
```diff
@@ -0,0 +1,345 @@
+	return templateItem, nil
+}
+
+// process the user provided values, create a device configuration and check if it is a valid device
```
// processDeviceValues ...

andig commented on cmd/configure/main.go:
```diff
@@ -0,0 +1,272 @@
+		}
+
+		chargingModes := []string{string(api.ModeOff), string(api.ModeNow), string(api.ModeMinPV), string(api.ModePV)}
+		ladeModi := []string{
```
englisch

DerAndereAndi committed:
Rename proxy files to template

DerAndereAndi committed:
Add comments to tests

DerAndereAndi committed:
Specifically allow yaml files in templates folders

DerAndereAndi committed:
Various review changes

DerAndereAndi committed:
Fix adding eebus loglevel

DerAndereAndi committed:
Add more comments

DerAndereAndi committed:
Rename consts into go style

DerAndereAndi commented on cmd/configure/devicetest.go:
```diff
@@ -0,0 +1,139 @@
+			return DeviceTestResult_Invalid, err
+		}
+	} else {
+		return DeviceTestResult_Invalid, errors.New("Selected device is not a wallbox!")
```
Eigentlich nicht, das sollte theoretisch nie vorkommen, da für den charger test ja nur charger auswählbar sind. Passiert also nur wenn ein Template grob falsch eingerichtet wurde.

DerAndereAndi replied:
Wobei, wenn ich ne falsche IP angebe, dann könnte das auch passieren. Ist bei allen Geräten dann so. Ich mache ne einheitliche übersetzte Meldung draus.

DerAndereAndi replied:
Also, Kommando zurück. Diese Meldungen wird der Anwender nie sehen. Das passiert nur wenn ein Gerät im Template z.b. als Meter angelegt ist, aber z.b. ein Charger ist.

DerAndereAndi commented on .gitignore:
```diff
@@ -4,7 +4,8 @@ __debug_bin
 *.py
 *.log
 *.json
-*.yaml
```
erledigt. schau bitte die Änderung dazu an

DerAndereAndi committed:
Update comments

DerAndereAndi commented on cmd/configure/configure.tpl:
```diff
@@ -0,0 +1,87 @@
+{{-     end }}
+{{-     if ne (len .Vehicles) 0 }}
+  vehicles:
+{{-       range .Vehicles }}
```
ich hab das lieber so wie es ist, da damit auch die logische Struktur/Hierarchie der If Abfragen etc. klar ist

DerAndereAndi committed:
Add more localized error texts
